// @phoenix/phnx-utils/lib/store/configureStore.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports._wrapWithTryCatch = _wrapWithTryCatch;
exports.getState = getState;
exports.getStore = getStore;
exports.replaceReducer = replaceReducer;
exports.dispatch = dispatch;
exports.default = exports.configUtils = void 0;
var configureStore;
var store;

if (process.env.NODE_ENV === 'production') {
  configureStore = require('./configureStore.prod');
} else {
  configureStore = require('./configureStore.dev');
}

function _wrapWithTryCatch(fn, identifier) {
  var ret;

  try {
    ret = fn();
  } catch (e) {
    throw new Error("Unable to ".concat(identifier, ", looks like the store has not been initialized, please initalize the store first"));
  }

  return ret;
}

function getState() {
  return _wrapWithTryCatch(function () {
    return store.getState();
  }, 'getState');
}

function getStore() {
  return store;
}

function replaceReducer(nextReducer) {
  return _wrapWithTryCatch(function () {
    return store.replaceReducer(nextReducer);
  }, 'replaceReducer');
}

function dispatch(payload) {
  if (store && store.dispatch) {
    return store.dispatch(payload);
  } else {
    throw new Error("Unable to dispatch, looks like the store has not been initialized, please initalize the store first");
  }
}

var configUtils = {
  getStore: getStore,
  getState: getState,
  replaceReducer: replaceReducer,
  dispatch: dispatch
};
exports.configUtils = configUtils;

function initialize(initialState) {
  store = configureStore.default(initialState);
  return store;
}

var _default = initialize;
exports.default = _default;
